### PR TITLE
SWAN IO (mhkit.wave.io.swan)

### PR DESCRIPTION
**Pull request by ssolson**
This pull request will add basic functionality to import table and block (ASCII and binary) output from SWAN into MHKiT. Functionality includes methods to convert block data into table data as demonstrated in the example notebook.

image

TODO:
 - [x] Pass Travis CI Test 
 - [x] Add netcdf block output support
 - [x] General Clean up

### CODE REVIEW

**Comment by rpauly18** [comment]
@ssolson One thing I have noticed right off the bat is that this PR contains commits from the open cdip PR. Does this functionality use that code/was that intentional?

**Comment by ssolson** [comment]
Good Catch! Removed the SWAN PR.

**Review by kmruehl — COMMENTED** [review]
@ssolson great job on this PR! It looks great. I made some minor revisions to the example, feel free to accept/reject them. I have a few inline comments, and then the following general comments:

1. Have you considered adding the SWAN plotting scripts from the example to the graphics toolbox? I think they are a great way to quickly plot SWAN data by calling a MHKiT function that is appropriately formatted for spatial data from from. 
2. Why do you need to convert the block data to a table to plot it? It seems like this would another great graphics function, to plot the block data as is, without conversion to a table. The block data is already formatted in such a way that it displays the spatial distribution of SWAN data, so plotting it directly makes sense to me.

File: examples/data/wave/swan/SWANOUTBlock.DAT
Comment (ssolson) [reply examples/data/wave/swan/SWANOUTBlock.DAT:5]: Hey Kelly that is correct. The uints are parsed and applied to the dataset if scaled.

File: mhkit/wave/graphics.py
Comment (ssolson) [reply mhkit/wave/graphics.py:387]: 🧠 Removed plot compendium.

File: mhkit/wave/graphics.py
Comment (ssolson) [reply mhkit/wave/graphics.py:459]: 🧠 Removed box plot..

File: mhkit/wave/graphics.py
Comment (ssolson) [reply mhkit/wave/graphics.py:590]: I have overwritten the graphics with the main repository function.

File: mhkit/wave/io/swan.py
Comment (ssolson) [reply mhkit/wave/io/swan.py:229]: I added a commit to sort block_to_table by xy pair. This will impact both of the functions you mention above as the dictionary of blocks is a wrapper around block_to_table.

**Comment by ssolson** [comment]
Thanks, Kelly I have addressed all of your code comments. Good suggestions and catches.

> 1. Have you considered adding the SWAN plotting scripts from the example to the graphics toolbox? I think they are a great way to quickly plot SWAN data by calling a MHKiT function that is appropriately formatted for spatial data from SWAN.

I have considered however I decided against it because there is not a standard method of SWAN output in terms of location (x,y) or naming convention of the variables. These vary based on output type and options. In my opinion, if there is not a standard output there is not a good way to write a generalized and meaningful plotting function. Essentially at best I see e.g. `plot_swan_table` being an almost verbatim wrapper around `tricontourf`. I am against writing such wrappers because it (IMO) adds little value to the user and dramatically increases our overhead (tests, maintenance, and documentation) down the road to ensure our wrapper functions are compatible with updates to basic matplotlib calls.

> 2. Why do you need to convert the block data to a table to plot it? It seems like this would another great graphics function, to plot the block data as is, without conversion to a table. The block data is already formatted in such a way that it displays the spatial distribution of SWAN data, so plotting it directly makes sense to me.

The primary reasoning for not having a dedicated function for this is the same as above. You do not have to convert to table to plot. I think the easiest method to get a contour from a block output would be imshow e.g. `plt.imshow(swan_block_mat['Hsig'])`  will give you a plot. I added an example to the end of the notebook for completeness but I do not think that we should write a wrapper around `imshow` for the same reasons I list above.

**Comment by kmruehl** [comment]
@ssolson I personally like having quick wrapper functions for graphics. It's a quick/easy want to visualize data, and that is what several of the graphics functions do already (e.g. `plot_sprectrum`), but will defer to you and @rpauly18. You may also consider having @ccchart review this PR.

**Comment by rpauly18** [comment]
I actually had a note to include suggesting making the graphics a function in MHKiT in my review, so I agree with Kelley.

**Comment by ssolson** [comment]
I disagree with this decision. The following is what is being proposed:
```
def plot_unstructured_contourf(x,y,z,label):
  '''
  Plots a filled contour for xyz data

  Parameters
  ------------
  x: array-like    
    X-data
  y: array-like    
    Y-data
  z: array-like    
    Z-data
  label: str
    label for contour

  Reuturns
  ---------
  ax: figure
    Unstructured contour
  '''

  plt.figure()
  ax = plt.tricontourf(x, y, z)
  cbar = plt.colorbar()
  cbar.set_label(label)
  return ax
```

There is nothing in this function about swan. This is a wrapper around tricontourf. This is bad coding practice for the maintenance, testing, and documentation reasons I listed above. Because there is nothing about swan here there is nothing related to MHK here and therefore is in my opinion not only bad coding practice but beyond the scope of our project. We have many things to add and I think wrapper functions around basic matplotlib calls adds little to no value. A user may write their own basic function to save themselves 3 lines of code but this is not the most effective use of the limited budget for me to go write this function and tests and then support changes and documentation for this down the line.

**Comment by rpauly18** [comment]
@ssolson I see what you are saying, but there may be ways we can make a plot such as this little more useful specifically to the MRE community. I will look into it. Regardless, we should add something to the SWAN example that uses MHKiT after the data has been ingested to provide a reason why users would want to use MHKiT to ingest the data. Many existing SWAN users likely already have their own code for reading in the data- so we need to give them incentive to switch to MHKiT.

**Review by rpauly18 — APPROVED** [review]
Great job on this. I made some minor changes to the naming conventions and asserts, reject if you do not like the changes. As for my previous comment about plots and an MHKiT example- if you cannot come up with anything to demonstrate, I would not hold up the PR, but it is something we should work on and add to the example as functions are done.

**Comment by ssolson** [comment]
Rebecca I was looking into the available functions to apply the significant wave height to a plot or something and the only thing I could see would be the Capture length matrix. For this, however, we would need to calculate the capture length. Do you have thoughts on applying these MHKiT functions?

Kelly has pointed to some spectrum data that the tutorial outputs which could also be added to the swam module but I would like to get this approved before adding any new input functions to the swam module.

**Comment by rpauly18** [comment]
I have no immediate ideas of what to include. You could generate fake power data or use WEC-Sim data for capture length. It would be good to figure out what kind of analysis people generally do with SWAN data and see if there are function gaps in the MHKiT wave module we can fill. I added some functions that I am working on to the roadmap that are used on needs people pointed out to me through discussing the WPTO hindcast data- you could do something similar. 

It would be good to add the additional functionality to the SWAN module, but no need to hold up this PR for that. It would be something good to add to the roadmap. I am also not trying to hold up this PR for the example notebook, but we should try and add to the example soon- and certainly before we highlight the swan module in any demonstrations.

**Comment by ssolson** [comment]
Hey Rebecca Kelly and I went over the functions today and without reading in new data we didn't see any additional feature to add. I changed the colormap of the plots to match (they differed between jet and Viridis before). I am ready to merge this if you are.

**Review by rpauly18 — APPROVED** [review]

**Pull request merged**
